archive/imo/imo1981_q3.lean

```diff
@@ -0,0 +1,235 @@
+problem_predicate N (n - m) m :=
+begin
+  have h3 : (m ^ 2 - (n - m) * m - (n - m) ^ 2) ^ 2 = (n ^ 2 - m * n - m ^ 2) ^ 2, by ring,
+  rw h1.right.right at h3,
+  obtain (h4|h4) : m = n ∨ m < n, from eq_or_lt_of_le (pp_m_le_n N h1),
+  { rw h4 at h1,
+    have h5 : n = 1, from pp_eq_imp_1 N h1,
+    exact absurd h5.symm (ne_of_lt h2) },
+  have h6 : 0 < n - m, from sub_pos.mpr h4,
+  have h7 : n - m < n, from sub_lt_self n h1.left.left,
+  have h8 : n ≤ N, from h1.right.left.right,
+  have h9 : n - m ≤ N, by linarith,
+  exact ⟨⟨h6, h9⟩, ⟨h1.1, h3⟩⟩
+end
+
+/-
+Now we can use induction to show that solutions must be Fibonacci numbers.
+It will be convenient to have the lemmas above in their natural number form.
+-/
+
+def nat_predicate (m n : ℕ) : Prop := problem_predicate N ↑m ↑n
+
+lemma np_m_le_n {m n : ℕ} (h1 : nat_predicate N m n) : m ≤ n :=
+begin
+  have h2 : ↑m ≤ ↑n, from pp_m_le_n N h1,
+  norm_cast at h2,
+  exact h2
+end
+
+lemma np_eq_imp_1 {n : ℕ} (h1 : nat_predicate N n n) : n = 1 :=
+begin
+  have h2 : ↑n = (1:ℤ), from pp_eq_imp_1 N h1,
+  norm_cast at h2,
+  exact h2
+end
+
+lemma np_reduction {m n : ℕ} (h1 : nat_predicate N m n) (h2 : 1 < n) :
+  nat_predicate N (n - m) m :=
+begin
+  have h3 : (1:ℤ) < ↑ n, by {norm_cast, exact h2},
+  unfold nat_predicate,
+  have h4 : m ≤ n, from np_m_le_n N h1,
+  have h5 : ↑(n - m) = ↑n - ↑m, from int.coe_nat_sub h4,
+  rw h5,
+  exact (pp_reduction N h1 h3)
+end
+
+lemma np_n_pos {m n : ℕ} (h1 : nat_predicate N m n) : 0 < n :=
+begin
+  have h2 : (0:ℤ) < ↑n, from h1.right.left.left,
+  norm_cast at h2,
+  exact h2
+end
+
+lemma np_m_pos {m n : ℕ} (h1 : nat_predicate N m n) : 0 < m :=
+by exact_mod_cast h1.1.1
+
+lemma np_n_le_N {m n : ℕ} (h1 : nat_predicate N m n) : n ≤ N :=
+by exact_mod_cast h1.2.1.2
+
+lemma np_imp_fib {n : ℕ} : ∀ m : ℕ, nat_predicate N m n →
+  ∃ k : ℕ, m = k.fib ∧ n = (k + 1).fib :=
+begin
+  apply nat.strong_induction_on n _,
+  intros n h1 m h2,
+  have h3 : 0 < n, from np_n_pos N h2,
+  have h4 : 0 < m, from np_m_pos N h2,
+  have h5 : m ≤ n, from np_m_le_n N h2,
+  obtain (h6|h6) : 1 = n ∨ 1 < n, from eq_or_lt_of_le (nat.succ_le_iff.mpr h3),
+  { have h7 : 1 ≤ m, from nat.succ_le_iff.mpr h4,
+    rw ← h6 at h5,
+    have h8 : m = 1, from le_antisymm h5 h7,
+    use 1,
+    simp only [nat.fib_one, nat.fib_two],
+    exact and.intro h8 h6.symm },
+  { have h9 : nat_predicate N (n - m) m, from np_reduction N h2 h6,
+    obtain (h10|h10) : m = n ∨ m < n, from eq_or_lt_of_le h5,
+    { rw h10 at h2,
+      exact absurd (np_eq_imp_1 N h2) (ne_of_gt h6) },
+    { obtain ⟨k, h11⟩ : ∃ k : ℕ, (n - m) = k.fib ∧ m = (k+1).fib, from h1 m h10 (n - m) h9,
+      use k + 1,
+      split,
+      { exact h11.right },
+      { ring,
+        rw [nat.fib_succ_succ, h11.left.symm, h11.right.symm],
+        exact (nat.sub_eq_iff_eq_add h5).mp rfl } } }
+end
+
+lemma nat_abs_coe (n : ℤ) (h1 : 0 < n) : n = ↑(n.nat_abs) :=
```

Comment:
This is `int.nat_abs_of_nonneg`

Reply:
aha, ok removed `nat_abs_coe` and rewrote to use `int.nat_abs_of_nonneg`